lib/ukdebug/gdbstub.c

```diff
@@ -0,0 +1,291 @@
+	else if ((hex >= 'A') && (hex <= 'F'))
+		return hex - 'A' + 0xa;
+
+	return -EINVAL;
+}
+
+static __ssz gdb_hex2mem(char *mem, __sz mem_len, const char *hex, __sz hex_len)
+{
+	__sz l = hex_len;
+	int i;
+
+	if (hex_len % 2 != 0)
+		return -EINVAL;
+
+	if (mem_len < hex_len / 2)
+		return -ENOMEM;
+
+	for (; l > 0; l -= 2, mem++) {
+		/* Decode high nibble */
+		i = gdb_hex2int(*hex++);
+		if (i < 0)
+			return i;
+
+		*mem = i << 4;
+
+		/* Decode low nibble */
+		i = gdb_hex2int(*hex++);
+		if (i < 0)
+			return i;
+
+		*mem |= i;
+	}
+
+	return hex_len / 2;
+}
+
+static __ssz gdb_send(const char *buf, __sz len)
+{
+	__sz l = len;
+
+	while (l--)
+		GDB_CHECK(gdb_dbg_putc(*buf++));
+
+	return len;
+}
+
+static __ssz gdb_recv(char *buf, __sz len)
+{
+	__sz l = len;
+	int r;
+
+	while (l--) {
+		r = gdb_dbg_getc();
+		if (r < 0)
+			return r;
+
+		*buf++ = (char)r;
+	}
+
+	return len;
+}
+
+static int gdb_send_ack(void)
+{
+	return gdb_dbg_putc('+');
+}
+
+static int gdb_send_nack(void)
+{
+	return gdb_dbg_putc('-');
+}
+
+static int gdb_recv_ack(void)
+{
+	int r = gdb_dbg_getc();
+
+	return (r < 0) ? r : ((char)r == '+');
+}
+
+static __ssz gdb_send_packet(const char *buf, __sz len)
+{
+	char hex[2];
+	char chksum = gdb_checksum(buf, len);
+	int r, retries = 0;
+
+	gdb_mem2hex(hex, sizeof(hex), &chksum, 1);
+
+	/* GDB packet format: $<DATA>#<CC>
+	 * where CC is the GDB packet checksum
+	 */
+	do {
+		if (retries++ > GDB_PACKET_RETRIES)
+			return -1;
+
+		GDB_CHECK(gdb_dbg_putc('$'));
+		GDB_CHECK(gdb_send(buf, len));
+		GDB_CHECK(gdb_dbg_putc('#'));
+		GDB_CHECK(gdb_send(hex, sizeof(hex)));
+	} while ((r = gdb_recv_ack()) == 0);
+
+	return (r == 1) ? (__ssz)len : r;
+}
+
+/* '' */
+static __ssz gdb_send_empty_packet(void)
+{
+	return gdb_send_packet(NULL, 0);
+}
+
+/* S nn */
+static __ssz gdb_send_signal_packet(int errnr)
+{
+	char buf[3];
+
+	UK_ASSERT((errnr & 0xff) == errnr);
+
+	buf[0] = 'S';
+	gdb_byte2hex(buf + 1, sizeof(buf) - 1, errnr);
+
+	return gdb_send_packet(buf, sizeof(buf));
+}
+
+/* E nn */
+static __ssz gdb_send_error_packet(int err)
+{
+	char buf[3];
+
+	UK_ASSERT((err & 0xff) == err);
+
+	buf[0] = 'E';
+	gdb_byte2hex(buf + 1, sizeof(buf) - 1, err);
+
+	return gdb_send_packet(buf, sizeof(buf));
+}
+
+static __ssz gdb_recv_packet(char *buf, __sz len)
+{
+	int c, retries = 0;
+	char *p = buf;
+	__sz n = 0;
+
+	/* Wait for packet start character */
+	while ((c = gdb_dbg_getc()) != '$') {
+		if (c < 0)
+			return c;
+	}
+
+	while (n < len) {
+		c = gdb_dbg_getc();
+		if (c < 0) {
+			return c;
+		} else if (c == '#') {
+			char hex[2];
+			char chksum = 0;
+
+			GDB_CHECK(gdb_recv(hex, sizeof(hex)));
+
+			if (gdb_hex2mem(&chksum, 1, hex, sizeof(hex)) < 0) {
+				GDB_CHECK(gdb_send_nack());
+				continue;
+			}
+
+			if (chksum != gdb_checksum(buf, n)) {
+				GDB_CHECK(gdb_send_nack());
+				continue;
+			}
+
+			GDB_CHECK(gdb_send_ack());
+
+			/* Null-terminate the data */
+			*p = 0;
+			return n;
+		} else if (c == '$') {
```

Comment:
Do you mean warning as in `uk_pr_warn` calls? We can't add print calls to the gdb stub because the gdb communication and the print output can go over the same serial connection. If we print something while the gdb host is waiting for a response to a packet, it will interpret whatever we print as the response. This leads to errors.